primo/opt_model/model_options.py

```diff
@@ -0,0 +1,310 @@
+        perform_clustering(wd, distance_threshold=10.0)
+
+        # Step 2: Identify list of wells belonging to each cluster
+        # Structure: {cluster_1: [index_1, index_2,..], cluster_2: [], ...}
+        col_names = wd.col_names
+        set_clusters = set(wd.data[col_names.cluster])
+        self.campaign_candidates = {
+            cluster: list(wd.data[wd.data[col_names.cluster] == cluster].index)
+            for cluster in set_clusters
+        }
+
+        # Step 3: Construct pairwise-metrics between wells in each cluster.
+        # Structure: {cluster: {(index_1, index_2): distance_12, ...}...}
+        self.pairwise_distance = self._pairwise_matrix(metric="distance")
+        self.pairwise_age_difference = self._pairwise_matrix(metric="age")
+        self.pairwise_depth_difference = self._pairwise_matrix(metric="depth")
+
+        # Construct owner well count data
+        operator_list = set(wd.data[col_names.operator_name])
+        self.owner_well_count = {owner: [] for owner in operator_list}
+        for well in wd:
+            # {Owner 1: [(c1, i2), (c1, i3), (c4, i7), ...], ...}
+            # Key => Owner name, Tuple[0] => cluster, Tuple[1] => index
+            self.owner_well_count[wd.data.loc[well, col_names.operator_name]].append(
+                (wd.data.loc[well, col_names.cluster], well)
+            )
+
+        # NOTE: Attributes _opt_model and _solver are defined in
+        # build_optimization_model and solve_model methods, respectively.
+        self._opt_model = None
+        self._solver = None
+        LOGGER.info("Finished optimization model inputs.")
+
+    @property
+    def get_total_budget(self):
+        """Returns scaled total budget [in million USD]"""
+        # Optimization model uses scaled total budget value to avoid numerical issues
+        return self.config.total_budget / 1e6
+
+    @property
+    def get_mobilization_cost(self):
+        """Returns scaled mobilization cost [in million USD]"""
+        # Optimization model uses Scaled mobilization costs to avoid numerical issues
+        return {
+            num_wells: cost / 1e6
+            for num_wells, cost in self.config.mobilization_cost.items()
+        }
+
+    @property
+    def get_max_cost_project(self):
+        """Returns scaled maximum cost of the project [in million USD]"""
+        if self.config.max_cost_project is None:
+            return None
+
+        return self.config.max_cost_project / 1e6
+
+    @property
+    def optimization_model(self):
+        """Returns the Pyomo optimization model"""
+        return self._opt_model
+
+    @property
+    def solver(self):
+        """Returns the solver object"""
+        return self._solver
+
+    def _pairwise_matrix(self, metric: str):
+        wd = self.config.well_data  # WellData object
+        # distance_matrix returns a numpy array.
+        metric_array = distance_matrix(wd, {metric: 1})
```

Comment:
@apd-ypuranik and @strahlw  Updated the method to use numpy methods. I feel that storing the entire matrix may require a lot of memory. That is why, I'm constructing the dict of dicts. Not sure if this is the best approach or not. We can improve it in the future.

Reply:
Agreed: we never meant full matrix but a vectorized calculation for pairwise distances on a cluster level. This works great I think :-)